---
title: '方法: HRESULT に例外を割り当てる'
description: COM メソッドから返された HRESULT 値を .NET メソッドによってスローされる例外にマップする方法を確認します。 ランタイムは、COM と .NET の間の遷移を処理します。
ms.date: 03/30/2017
dev_langs:
- cpp
helpviewer_keywords:
- interoperation with unmanaged code, HRESULTs
- exceptions, HRESULTs
- interoperation with unmanaged code, exceptions
- HRESULTs
- COM interop, HRESULTs
- COM interop, exceptions
ms.assetid: 610b364b-2761-429d-9c4a-afbc3e66f1b9
ms.openlocfilehash: 13798de1547428d54c10f83c41a5a402e2b3fb53
ms.sourcegitcommit: c7f0beaa2bd66ebca86362ca17d673f7e8256ca6
ms.translationtype: HT
ms.contentlocale: ja-JP
ms.lasthandoff: 03/23/2021
ms.locfileid: "104872406"
---
# <a name="how-to-map-hresults-and-exceptions"></a><span data-ttu-id="8d965-104">方法: HRESULT に例外を割り当てる</span><span class="sxs-lookup"><span data-stu-id="8d965-104">How to: Map HRESULTs and Exceptions</span></span>

<span data-ttu-id="8d965-105">COM メソッドでは、HRESULT を返してエラーを報告します。 .NET メソッドでは、例外をスローしてエラーを報告します。</span><span class="sxs-lookup"><span data-stu-id="8d965-105">COM methods report errors by returning HRESULTs; .NET methods report them by throwing exceptions.</span></span> <span data-ttu-id="8d965-106">ランタイムは、この 2 つの間の遷移を処理します。</span><span class="sxs-lookup"><span data-stu-id="8d965-106">The runtime handles the transition between the two.</span></span> <span data-ttu-id="8d965-107">.NET Framework の例外クラスはそれぞれ HRESULT に割り当てられます。</span><span class="sxs-lookup"><span data-stu-id="8d965-107">Each exception class in the .NET Framework maps to an HRESULT.</span></span>

 <span data-ttu-id="8d965-108">ユーザー定義の例外クラスは、適切な HRESULT であればどの HRESULT でも指定できます。</span><span class="sxs-lookup"><span data-stu-id="8d965-108">User-defined exception classes can specify whatever HRESULT is appropriate.</span></span> <span data-ttu-id="8d965-109">これらの例外クラスでは、例外オブジェクトの `HResult` フィールドを設定することで例外が生成されたときに返される HRESULT を動的に変更できます。</span><span class="sxs-lookup"><span data-stu-id="8d965-109">These exception classes can dynamically change the HRESULT to be returned when the exception is generated by setting the `HResult` field on the exception object.</span></span> <span data-ttu-id="8d965-110">アンマネージ プロセスの .NET オブジェクトに実装されている `IErrorInfo` インターフェイスを通じて、クライアントに例外についての追加情報が提供されます。</span><span class="sxs-lookup"><span data-stu-id="8d965-110">Additional information about the exception is provided to the client through the `IErrorInfo` interface, which is implemented on the .NET object in the unmanaged process.</span></span>

 <span data-ttu-id="8d965-111">`System.Exception` を拡張するクラスを作成する場合は、構築中に HRESULT フィールドを設定する必要があります。</span><span class="sxs-lookup"><span data-stu-id="8d965-111">If you create a class that extends `System.Exception`, you must set the HRESULT field during construction.</span></span> <span data-ttu-id="8d965-112">それ以外の場合は、基本クラスで HRESULT 値が割り当てられます。</span><span class="sxs-lookup"><span data-stu-id="8d965-112">Otherwise, the base class assigns the HRESULT value.</span></span> <span data-ttu-id="8d965-113">例外のコンストラクターで値を指定することで、既存の HRESULT に新しい例外クラスを割り当てることができます。</span><span class="sxs-lookup"><span data-stu-id="8d965-113">You can map new exception classes to an existing HRESULT by supplying the value in the exception's constructor.</span></span>

 <span data-ttu-id="8d965-114">スレッドに `IErrorInfo` がある場合、ランタイムが `HRESULT` を無視することがあることに注意してください。</span><span class="sxs-lookup"><span data-stu-id="8d965-114">Note that the runtime will sometimes ignore an `HRESULT` in cases where there is an `IErrorInfo` present on the thread.</span></span>  <span data-ttu-id="8d965-115">この動作は、`HRESULT` と `IErrorInfo` が同じエラーを表していない場合に発生する可能性があります。</span><span class="sxs-lookup"><span data-stu-id="8d965-115">This behavior can occur in cases where the `HRESULT` and the `IErrorInfo` do not represent the same error.</span></span>

### <a name="to-create-a-new-exception-class-and-map-it-to-an-hresult"></a><span data-ttu-id="8d965-116">新しい例外クラスを作成し、HRESULT に割り当てるには</span><span class="sxs-lookup"><span data-stu-id="8d965-116">To create a new exception class and map it to an HRESULT</span></span>

1. <span data-ttu-id="8d965-117">次のコードを使用して、`NoAccessException` という新しい例外クラスを作成し、HRESULT `E_ACCESSDENIED` に割り当てます。</span><span class="sxs-lookup"><span data-stu-id="8d965-117">Use the following code to create a new exception class called `NoAccessException` and map it to the HRESULT `E_ACCESSDENIED`.</span></span>

    ```cpp
    Class NoAccessException : public ApplicationException
    {
        NoAccessException () {
        HResult = E_ACCESSDENIED;
    }
    }
    CMyClass::MethodThatThrows
    {
    throw new NoAccessException();
    }
    ```

 <span data-ttu-id="8d965-118">マネージド コードとアンマネージド コードの両方を同時に使用する (任意のプログラミング言語の) プログラムもあります。</span><span class="sxs-lookup"><span data-stu-id="8d965-118">You might encounter a program (in any programming language) that uses both managed and unmanaged code at the same time.</span></span> <span data-ttu-id="8d965-119">たとえば、次のコード例のカスタム マーシャラーは、`Marshal.ThrowExceptionForHR(int HResult)` メソッドを使用して、特定の HRESULT 値を持つ例外をスローします。</span><span class="sxs-lookup"><span data-stu-id="8d965-119">For example, the custom marshaler in the following code example uses the `Marshal.ThrowExceptionForHR(int HResult)` method to throw an exception with a specific HRESULT value.</span></span> <span data-ttu-id="8d965-120">このメソッドは、HRESULT を調べ、適切な例外型を生成します。</span><span class="sxs-lookup"><span data-stu-id="8d965-120">The method looks up the HRESULT and generates the appropriate exception type.</span></span> <span data-ttu-id="8d965-121">たとえば、次のコード フラグメントの HRESULT では、`ArgumentException` が生成されます。</span><span class="sxs-lookup"><span data-stu-id="8d965-121">For example, the HRESULT in the following code fragment generates `ArgumentException`.</span></span>

```cpp
CMyClass::MethodThatThrows
{
    Marshal.ThrowExceptionForHR(COR_E_ARGUMENT);
}
```

 <span data-ttu-id="8d965-122">次の表には、HRESULT から .NET の同等の例外クラスへの一般的なマッピングが示されています。</span><span class="sxs-lookup"><span data-stu-id="8d965-122">The following table provides the common mappings from HRESULT to its comparable exception class in .NET.</span></span> <span data-ttu-id="8d965-123">明示的なマッピングのない HRESULT 値は、`COMException` にマップされます。</span><span class="sxs-lookup"><span data-stu-id="8d965-123">HRESULT values without explicit mappings are mapped to `COMException`.</span></span> <span data-ttu-id="8d965-124">完全な最新のマッピングについては、[dotnet/runtime リポジトリ](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/rexcep.h)を参照してください。</span><span class="sxs-lookup"><span data-stu-id="8d965-124">The complete up-to-date mapping can be found in the [dotnet/runtime repository](https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/rexcep.h).</span></span>

|<span data-ttu-id="8d965-125">HRESULT</span><span class="sxs-lookup"><span data-stu-id="8d965-125">HRESULT</span></span>|<span data-ttu-id="8d965-126">.NET の例外</span><span class="sxs-lookup"><span data-stu-id="8d965-126">.NET exception</span></span>|
|-------------|--------------------|
|`COR_E_APPLICATION`|`ApplicationException`|
|<span data-ttu-id="8d965-127">`COR_E_ARGUMENT` または `E_INVALIDARG`</span><span class="sxs-lookup"><span data-stu-id="8d965-127">`COR_E_ARGUMENT` or `E_INVALIDARG`</span></span>|`ArgumentException`|
|`COR_E_ARGUMENTOUTOFRANGE`|`ArgumentOutOfRangeException`|
|`COR_E_ARITHMETIC or ERROR_ARITHMETIC_OVERFLOW`|`ArithmeticException`|
|`COR_E_ARRAYTYPEMISMATCH`|`ArrayTypeMismatchException`|
|`COR_E_BADIMAGEFORMAT or ERROR_BAD_FORMAT`|`BadImageFormatException`|
|`COR_E_DIRECTORYNOTFOUND or ERROR_PATH_NOT_FOUND`|`DirectoryNotFoundException`|
|`COR_E_DIVIDEBYZERO`|`DivideByZeroException`|
|`COR_E_DUPLICATEWAITOBJECT`|`DuplicateWaitObjectException`|
|`COR_E_ENDOFSTREAM`|`EndOfStreamException`|
|`COR_E_ENTRYPOINTNOTFOUND`|`EntryPointNotFoundException`|
|`COR_E_EXCEPTION`|`Exception`|
|`COR_E_EXECUTIONENGINE`|`ExecutionEngineException`|
|`COR_E_FIELDACCESS`|`FieldAccessException`|
|`COR_E_FILENOTFOUND or ERROR_FILE_NOT_FOUND`|`FileNotFoundException`|
|`COR_E_FORMAT`|`FormatException`|
|`COR_E_INDEXOUTOFRANGE`|`IndexOutOfRangeException`|
|`COR_E_INVALIDCAST or E_NOINTERFACE`|`InvalidCastException`|
|`COR_E_INVALIDFILTERCRITERIA`|`InvalidFilterCriteriaException`|
|`COR_E_INVALIDOPERATION`|`InvalidOperationException`|
|`COR_E_IO`|`IOException`|
|`COR_E_MEMBERACCESS`|`AccessException`|
|`COR_E_METHODACCESS`|`MethodAccessException`|
|`COR_E_MISSINGFIELD`|`MissingFieldException`|
|`COR_E_MISSINGMANIFESTRESOURCE`|`MissingManifestResourceException`|
|`COR_E_MISSINGMEMBER`|`MissingMemberException`|
|`COR_E_MISSINGMETHOD`|`MissingMethodException`|
|`COR_E_NOTFINITENUMBER`|`NotFiniteNumberException`|
|`E_NOTIMPL`|`NotImplementedException`|
|`COR_E_NOTSUPPORTED`|`NotSupportedException`|
|`COR_E_NULLREFERENCE orE_POINTER`|`NullReferenceException`|
|`COR_E_OUTOFMEMORY or`<br /><br /> `E_OUTOFMEMORY`|`OutOfMemoryException`|
|`COR_E_OVERFLOW`|`OverflowException`|
|`COR_E_PATHTOOLONG or ERROR_FILENAME_EXCED_RANGE`|`PathTooLongException`|
|`COR_E_RANK`|`RankException`|
|`COR_E_REFLECTIONTYPELOAD`|`ReflectionTypeLoadException`|
|`COR_E_SECURITY`|`SecurityException`|
|`COR_E_SERIALIZATION`|`SerializationException`|
|`COR_E_STACKOVERFLOW orERROR_STACK_OVERFLOW`|`StackOverflowException`|
|`COR_E_SYNCHRONIZATIONLOCK`|`SynchronizationLockException`|
|`COR_E_SYSTEM`|`SystemException`|
|`COR_E_TARGET`|`TargetException`|
|`COR_E_TARGETINVOCATION`|`TargetInvocationException`|
|`COR_E_TARGETPARAMCOUNT`|`TargetParameterCountException`|
|`COR_E_THREADINTERRUPTED`|`ThreadInterruptedException`|
|`COR_E_THREADSTATE`|`ThreadStateException`|
|`COR_E_TYPELOAD`|`TypeLoadException`|
|`COR_E_TYPEINITIALIZATION`|`TypeInitializationException`|
|`COR_E_VERIFICATION`|`VerificationException`|

 <span data-ttu-id="8d965-128">拡張エラー情報を取得するために、マネージド クライアントは、生成された例外オブジェクトのフィールドを調べる必要があります。</span><span class="sxs-lookup"><span data-stu-id="8d965-128">To retrieve extended error information, the managed client must examine the fields of the exception object that was generated.</span></span> <span data-ttu-id="8d965-129">例外オブジェクトでエラーについての有益な情報を提供できるようにするには、COM オブジェクトは、`IErrorInfo` インターフェイスを実装している必要があります。</span><span class="sxs-lookup"><span data-stu-id="8d965-129">For the exception object to provide useful information about an error, the COM object must implement the `IErrorInfo` interface.</span></span> <span data-ttu-id="8d965-130">ランタイムは、`IErrorInfo` で提供される情報を使用して例外オブジェクトを初期化します。</span><span class="sxs-lookup"><span data-stu-id="8d965-130">The runtime uses the information provided by `IErrorInfo` to initialize the exception object.</span></span>

 <span data-ttu-id="8d965-131">COM オブジェクトで `IErrorInfo` がサポートされていない場合、ランタイムは既定値で例外オブジェクトを初期化します。</span><span class="sxs-lookup"><span data-stu-id="8d965-131">If the COM object does not support `IErrorInfo`, the runtime initializes an exception object with default values.</span></span> <span data-ttu-id="8d965-132">次の表で、例外オブジェクトに関連付けられた各フィールドをリストし、COM オブジェクトが `IErrorInfo` をサポートしている場合の既定の情報ソースを示します。</span><span class="sxs-lookup"><span data-stu-id="8d965-132">The following table lists each field associated with an exception object and identifies the source of default information when the COM object supports `IErrorInfo`.</span></span>

 <span data-ttu-id="8d965-133">スレッドに `IErrorInfo` がある場合、ランタイムが `HRESULT` を無視することがあることに注意してください。</span><span class="sxs-lookup"><span data-stu-id="8d965-133">Note that the runtime will sometimes ignore an `HRESULT` in cases where there is an `IErrorInfo` present on the thread.</span></span>  <span data-ttu-id="8d965-134">この動作は、`HRESULT` と `IErrorInfo` が同じエラーを表していない場合に発生する可能性があります。</span><span class="sxs-lookup"><span data-stu-id="8d965-134">This behavior can occur in cases where the `HRESULT` and the `IErrorInfo` do not represent the same error.</span></span>

|<span data-ttu-id="8d965-135">例外フィールド</span><span class="sxs-lookup"><span data-stu-id="8d965-135">Exception field</span></span>|<span data-ttu-id="8d965-136">COM からの情報のソース</span><span class="sxs-lookup"><span data-stu-id="8d965-136">Source of Information from COM</span></span>|
|---------------------|------------------------------------|
|`ErrorCode`|<span data-ttu-id="8d965-137">呼び出しから返された HRESULT。</span><span class="sxs-lookup"><span data-stu-id="8d965-137">HRESULT returned from call.</span></span>|
|`HelpLink`|<span data-ttu-id="8d965-138">`IErrorInfo->HelpContext` が 0 以外の場合、文字列は `IErrorInfo->GetHelpFile`、"#"、および `IErrorInfo->GetHelpContext` を連結して形成されます。</span><span class="sxs-lookup"><span data-stu-id="8d965-138">If `IErrorInfo->HelpContext` is nonzero, the string is formed by concatenating `IErrorInfo->GetHelpFile` and "#" and `IErrorInfo->GetHelpContext`.</span></span> <span data-ttu-id="8d965-139">それ以外の場合、文字列は `IErrorInfo->GetHelpFile` から返されます。</span><span class="sxs-lookup"><span data-stu-id="8d965-139">Otherwise the string is returned from `IErrorInfo->GetHelpFile`.</span></span>|
|`InnerException`|<span data-ttu-id="8d965-140">常に null 参照 (Visual Basic では `Nothing`)。</span><span class="sxs-lookup"><span data-stu-id="8d965-140">Always a null reference (`Nothing` in Visual Basic).</span></span>|
|`Message`|<span data-ttu-id="8d965-141">`IErrorInfo->GetDescription` から返される文字列。</span><span class="sxs-lookup"><span data-stu-id="8d965-141">String returned from `IErrorInfo->GetDescription`.</span></span>|
|`Source`|<span data-ttu-id="8d965-142">`IErrorInfo->GetSource` から返される文字列。</span><span class="sxs-lookup"><span data-stu-id="8d965-142">String returned from `IErrorInfo->GetSource`.</span></span>|
|`StackTrace`|<span data-ttu-id="8d965-143">スタック トレース。</span><span class="sxs-lookup"><span data-stu-id="8d965-143">The stack trace.</span></span>|
|`TargetSite`|<span data-ttu-id="8d965-144">失敗を示す HRESULT を返したメソッドの名前。</span><span class="sxs-lookup"><span data-stu-id="8d965-144">The name of the method that returned the failing HRESULT.</span></span>|

 <span data-ttu-id="8d965-145">`Message`、`Source`、および `StackTrace` などの例外フィールドは、`StackOverflowException` では使用できません。</span><span class="sxs-lookup"><span data-stu-id="8d965-145">Exception fields, such as `Message`, `Source`, and `StackTrace` are not available for the `StackOverflowException`.</span></span>

## <a name="see-also"></a><span data-ttu-id="8d965-146">関連項目</span><span class="sxs-lookup"><span data-stu-id="8d965-146">See also</span></span>

- <span data-ttu-id="8d965-147">[高度な COM 相互運用性](/previous-versions/dotnet/netframework-4.0/bd9cdfyx(v=vs.100))</span><span class="sxs-lookup"><span data-stu-id="8d965-147">[Advanced COM Interoperability](/previous-versions/dotnet/netframework-4.0/bd9cdfyx(v=vs.100))</span></span>
- [<span data-ttu-id="8d965-148">例外</span><span class="sxs-lookup"><span data-stu-id="8d965-148">Exceptions</span></span>](../../standard/exceptions/index.md)
